ros/angel_system_nodes/angel_system_nodes/task_monitor_v2.py

```diff
@@ -0,0 +1,282 @@
+        """
+        HMM runtime thread that gets the current state.
+        Thread is awakened each time a new ActivityDetection message is
+        received.
+
+        It is expected that the HMM's current state computation will finish
+        well before a new ActivityDetection message arrives, but we put it in a
+        separate thread just to be safe.
+        """
+        log = self.get_logger()
+        log.info("HMM state thread started")
+
+        while self._hmm_active.wait(0): # will quickly return false if cleared
+            if self._hmm_state_awake_evt.wait(self._hmm_active_heartbeat):
+                log.info("HMM state loop awakened")
+                self._hmm_state_awake_evt.clear()
+
+                # Get the HMM prediction
+                start_time = time.time()
+                with self._hmm_lock:
+                    step = self._hmm.get_current_state()
+                log.info(f"HMM get_current_state time: {time.time() - start_time}")
+
+                if self._current_step != step:
+                    self._previous_step = self._current_step
+                    self._current_step = step
+
+                log.info(f"Current step {self._current_step}")
+
+                # Publish a new TaskUpdate message
+                self.publish_task_state_message()
+
+    def thread_run_hmm_skip(self):
+        """
+        HMM runtime thread that gets the current likelihood of a step skip.
+        Thread is awakened each time a new ActivityDetection message is
+        received.
+
+        It is expected that the skip score computation will finish well before
+        a new ActivityDetection message arrives, but we put it in a separate
+        thread just to be safe.
+        """
+        log = self.get_logger()
+        log.info("HMM skip thread started")
+
+        while self._hmm_active.wait(0): # will quickly return false if cleared
+            if self._hmm_skip_awake_evt.wait(self._hmm_active_heartbeat):
+                log.info("HMM skip loop awakened")
+                self._hmm_skip_awake_evt.clear()
+
+                # Get the HMM prediction
+                start_time = time.time()
+                with self._hmm_lock:
+                    skip_score = self._hmm.get_skip_score()
+                log.info(f"HMM get_skip_score time: {time.time() - start_time}")
+                log.info(f"Skip score: {skip_score}")
+
+                # TODO: how does skip score relate to task completion conf.
+                self._task_complete_confidence = skip_score
+
+    def hmm_alive(self) -> bool:
+        """
+        Check that the HMM runtime is still alive and raise an exception
+        if it is not.
+        """
+        alive = (
+            self._hmm_state_thread.is_alive() and self._hmm_skip_thread.is_alive()
+        )
+        if not alive:
+            self.get_logger().warn("HMM thread no longer alive.")
+            self._hmm_active.clear()  # make HMM active flag "False"
```

Comment:
both events need to be cleared because either may have triggered entering this block.